File: ObjCSample/ObjCSample/SampleControllers/Multiple Banners in UITableView/BannerInTableViewController.m
//
//  BannerInTableViewController.m
//  ObjCSample
//
//  Created by Julien Gomez on 17/10/2018.
//  Copyright © 2019 Smart AdServer. All rights reserved.
//

#import "BannerInTableViewController.h"

#define kSiteID             104808
#define kBanner1PageID      663262
#define kBanner2PageID      663531
#define kBanner3PageID      663530
#define kFormatID           15140

#define kBanner1Position    6
#define kBanner2Position    12
#define kBanner3Position    18
#define kNumberOfCells      40

#define kDefaultCellHeight  80

#define kDescriptionCellIdentifier  @"DescriptionCell"
#define kContentCellIdentifier      @"ContentCell"

@interface BannerInTableViewController ()

@property (nonatomic, strong) CustomBannerView *banner1;
@property (nonatomic, strong) CustomBannerView *banner2;
@property (nonatomic, strong) CustomBannerView *banner3;

@property (nonatomic, strong) UIView *stickyAdView;

@property (nonatomic, assign) CGFloat navigationBarTopOffset;
@property (nonatomic, assign) BOOL isObservingKeyPathHidden;

@end

@implementation BannerInTableViewController

/* WARNING:
 * Note that in this class, a lot of code in the following methods…
 * - (void)viewWillAppear:(BOOL)animated
 * - (void)viewWillDisappear:(BOOL)animated
 * - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
 *
 * …is only here to demonstrate how to handle "Stick to top" ads. You can skip this part if you don't plan to display such ads.
 */

#pragma mark - Object lifecycle

- (void)dealloc {
    [self removeBanners];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Initialize the refresh control.
    self.refreshControl = [[UIRefreshControl alloc] init];
    [self.refreshControl addTarget:self action:@selector(createBanners) forControlEvents:UIControlEventValueChanged];
    
    // Create the banners
    [self createBanners];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    // Hide navigation bar on swipe gesture
    self.navigationController.hidesBarsOnSwipe = YES;
    
    // Add observer for hidden state changed. This will be usefull to animate "StickToTop" Adviews.
    if (!self.isObservingKeyPathHidden) {
        [self.navigationController.navigationBar addObserver:self forKeyPath:@"hidden" options:NSKeyValueObservingOptionNew context:NULL];
        self.isObservingKeyPathHidden = YES;
    }
    
    // Get initial top offset of the navigation bar
    self.navigationBarTopOffset = self.navigationController.navigationBar.frame.origin.y;
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    self.navigationController.hidesBarsOnSwipe = NO;
    [self.navigationController setNavigationBarHidden:NO animated:YES];
    
    // Remove key value observing
    if (self.isObservingKeyPathHidden) {
        [self.navigationController.navigationBar removeObserver:self forKeyPath:@"hidden"];
        self.isObservingKeyPathHidden = NO;
    }
}

- (void)createBanners {
    self.banner1 = [self createBanner:kBanner1PageID];
    self.banner2 = [self createBanner:kBanner2PageID];
    self.banner3 = [self createBanner:kBanner3PageID];
}

- (void)removeBanners {
    [self removeBanner:self.banner1];
    [self removeBanner:self.banner2];
    [self removeBanner:self.banner3];
}

- (CustomBannerView *)createBanner:(unsigned long)pageId {
    CustomBannerView *banner = [[CustomBannerView alloc] initWithFrame:CGRectMake(0, 0, CGRectGetWidth(self.view.frame), kDefaultCellHeight)];
    banner.delegate = self;
    banner.modalParentViewController = self;
    
    // To avoid having the parallax going below your navigation bar and your action bar, you can manually set the parallax infos.
    // Uncomment the code below if you want to manually set the viewport of your parallax.
    //
    // CGFloat topBarSize = YOUR_TOP_BAR_SIZE;
    // CGFloat bottomBarSize = YOUR_BOTTOM_BAR_SIZE;
    // CGFloat parallaxViewportHeight = [UIScreen mainScreen].bounds.size.height - topBarSize - bottomBarSize;
    // [banner setParallaxInfos:[[SASParallaxInfos alloc] initWithViewportTopOrigin:topBarSize viewportHeight:parallaxViewportHeight]];
    
    [banner loadWithPlacement:[SASAdPlacement adPlacementWithSiteId:kSiteID pageId:pageId formatId:kFormatID]];
    
    return banner;
}

- (void)removeBanner:(CustomBannerView *) banner {
    [banner removeFromSuperview];
    banner.modalParentViewController = nil;
    banner.delegate = nil;
    banner.loaded = NO;
}

#pragma mark - Table view data source

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return kNumberOfCells;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (kBanner1Position == indexPath.row) {
        return [self.banner1 optimalAdViewHeightForContainer:tableView];
    } else if (kBanner2Position == indexPath.row) {
        return [self.banner2 optimalAdViewHeightForContainer:tableView];
    } else if (kBanner3Position == indexPath.row) {
        return [self.banner3 optimalAdViewHeightForContainer:tableView];
    } else {
        return kDefaultCellHeight;
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (kBanner1Position == indexPath.row) {
        return [SASAdViewContainerCell cellForAdView:self.banner1 inTableView:tableView];
    } else if (kBanner2Position == indexPath.row) {
        return [SASAdViewContainerCell cellForAdView:self.banner2 inTableView:tableView];
    } else if (kBanner3Position == indexPath.row) {
        return [SASAdViewContainerCell cellForAdView:self.banner3 inTableView:tableView];
    } else if (indexPath.row == 0) {
        // Configure description cell
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kDescriptionCellIdentifier];
        return cell;
    } else {
        // Configure content cell
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kContentCellIdentifier];

        UILabel *indexLabel = (UILabel *)[cell viewWithTag:1];
        indexLabel.text = [NSString stringWithFormat:@"%ld", (long)indexPath.row];
        indexLabel.layer.masksToBounds = true;
        indexLabel.layer.cornerRadius = 15;
        cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
        return cell;
    }
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    // Deselect the cell if it's an ad to avoid messing with the ad creative.
    if (kBanner1Position == indexPath.row || kBanner2Position == indexPath.row || kBanner3Position == indexPath.row) {
        // Ad Cell was selected, nothing to do here.
    } else {
        // Your normal behavior on cell click like pushing a new VC
        [self pushNewViewController];
    }
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
}

- (void)pushNewViewController {
    UIViewController *controller = [[UIViewController alloc] init];
    controller.navigationItem.title = @"Details";
    controller.view.backgroundColor = self.view.backgroundColor;
    UILabel *label = [[UILabel alloc] initWithFrame:controller.view.bounds];
    label.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    label.textAlignment = NSTextAlignmentCenter;
    label.text = @"Detail view controller";
    [controller.view addSubview:label];
    [self.navigationController pushViewController:controller animated:YES];
}

#pragma mark - Observe navigation bar hidden state to animate sticky view if needed

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    if ([keyPath isEqualToString:@"hidden"] && self.stickyAdView) {
        UINavigationBar *navigationBar = self.navigationController.navigationBar;
        if (navigationBar.hidden == NO) {
            [UIView animateWithDuration:0.3 animations:^{
                CGRect newFrame = self.stickyAdView.frame;
                newFrame.origin.y = self.navigationBarTopOffset + CGRectGetHeight(navigationBar.frame);
                self.stickyAdView.frame = newFrame;
            }];
        } else {
            [UIView animateWithDuration:0.3 animations:^{
                CGRect newFrame = self.stickyAdView.frame;
                newFrame.origin.y = self.navigationBarTopOffset;
                self.stickyAdView.frame = newFrame;
            }];
        }
    }
}

#pragma mark - SASBannerView delegate

- (void)bannerView:(SASBannerView *)bannerView didDownloadAd:(SASAd *)ad {
    [self.tableView reloadData];
}

- (void)bannerViewDidLoad:(SASBannerView *)bannerView {
    NSLog(@"A Banner has been loaded");
    CustomBannerView *banner = (CustomBannerView *)bannerView;
    banner.loaded = YES;

    // Hide refresh control when all banners are fully loaded
    if (self.banner1.loaded && self.banner2.loaded && self.banner3.loaded) {
        [self.refreshControl endRefreshing];
    }
}

- (void)bannerView:(SASBannerView *)bannerView didFailToLoadWithError:(NSError *)error {
    NSLog(@"Banner has failed to load with error: %@", [error description]);
    CustomBannerView *banner = (CustomBannerView *)bannerView;
    banner.loaded = YES;
    
    // Hide refresh control when banners are fully loaded
    if (self.banner1.loaded && self.banner2.loaded && self.banner3.loaded) {
        [self.refreshControl endRefreshing];
    }
}

- (void)bannerView:(SASBannerView *)bannerView didSendVideoEvent:(SASVideoEvent)videoEvent {
    // This delegate can be used to listen for events if the ad is a video.
    
    // For instance, you can use these events to check if the video has been played until the end
    // by listening to the event 'SASVideoEventComplete'
    
    if (videoEvent == SASVideoEventComplete) {
        NSLog(@"The video has been played until the end");
    }
}

- (void)bannerView:(SASBannerView *)bannerView withStickyView:(UIView *)stickyView didStick:(BOOL)stuck withFrame:(CGRect)stickyFrame {
    // This delegate can be used to be notified when a video read banner did stick or unstick, in this implementation as the navigation bar appears / disappears during vertical swipe we need to animate the sticky view

    if (stuck) {
        // Keep a reference to the view to animate it if necessary
        self.stickyAdView = stickyView;
    } else {
        self.stickyAdView = nil;
    }
}

- (BOOL)bannerViewCanStickToTop:(SASBannerView *)bannerView {
    return YES;
}

@end

File: SwiftSample/SwiftSample/SampleControllers/Multiple Banners in UITableView/BannerInTableViewController.swift
//
//  BannerInTableViewController.swift
//  SwiftSample
//
//  Created by Julien Gomez on 17/10/2018.
//  Copyright © 2019 Smart AdServer. All rights reserved.
//

import AVFoundation
import UIKit
import SASDisplayKit

/*
 The purpose of this view controller is to display multiple banners in a table view.
 
 Displaying banners with various format and size in a table view that can hide its navigation
 bar can be tricky, this is an example on how it can be done.
 */
class BannerInTableViewController: UITableViewController, SASBannerViewDelegate {
    
    fileprivate let banner1Row = 6
    fileprivate let banner2Row = 12
    fileprivate let banner3Row = 18
    fileprivate let numberOfCells = 40
    
    fileprivate let defaultCellHeight: CGFloat = 80.0
    
    // Instances of banners (marked as optional since it is created after the initialization of the controller)
    fileprivate var banner1: SASBannerView?
    fileprivate var banner2: SASBannerView?
    fileprivate var banner3: SASBannerView?
    
    fileprivate var statusBarHidden = false
    
    fileprivate var stickyAdView: UIView?
    fileprivate var navigationBarTopOffset: CGFloat = 0.0
    
    fileprivate var isObserverRegistered: Bool = false
    
    // MARK: - View controller lifecycle
    
    deinit {
        removeBanners()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let refreshControl = UIRefreshControl()
        refreshControl.addTarget(self, action: #selector(BannerInTableViewController.handleRefresh(_:)), for: UIControl.Event.valueChanged)
        self.refreshControl = refreshControl
        
        createBanners()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if let navigationController = self.navigationController {
            // Hide navigation bar on swipe gesture
            navigationController.hidesBarsOnSwipe = true
            // Add observer for hidden state changed
            navigationController.navigationBar.addObserver(self, forKeyPath: "hidden", options: [], context: nil)
            isObserverRegistered = true
            // Get initial top offset of the navigation bar
            navigationBarTopOffset = navigationController.navigationBar.frame.origin.y;
        }
        
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        
        if let navigationController = self.navigationController {
            navigationController.hidesBarsOnSwipe = false
            navigationController.setNavigationBarHidden(false, animated: true)
            if isObserverRegistered == true {
                navigationController.navigationBar.removeObserver(self, forKeyPath: "hidden")
                isObserverRegistered = false
            }
        }
    }
    
    override var prefersStatusBarHidden : Bool {
        // The status bar can be hidden by an ad (for instance when the ad is expanded in fullscreen)
        return statusBarHidden || super.prefersStatusBarHidden
    }
    
    // MARK: - Ad management
    
    @objc func handleRefresh(_ refreshControl: UIRefreshControl) {
        self.banner1?.refresh()
        self.banner2?.refresh()
        self.banner3?.refresh()
        self.refreshControl?.endRefreshing()
    }
    
    func createBanners() {
        self.banner1 = createBanner(Constants.banner1InTableViewPageID)
        self.banner2 = createBanner(Constants.banner2InTableViewPageID)
        self.banner3 = createBanner(Constants.banner3InTableViewPageID)
    }
    
    func removeBanners() {
        removeBanner(self.banner1)
        removeBanner(self.banner2)
        removeBanner(self.banner3)
    }
    
    func createBanner(_ pageId: Int) -> SASBannerView {
        // The instance of the banner is created with a default frame and an appropriate loader.
        let banner = SASBannerView(frame: bannerFrameForView(self.view, height: defaultCellHeight), loader: .activityIndicatorStyleWhite)
        
        // Setting the delegate.
        banner.delegate = self
        
        // Setting the modal parent view controller.
        banner.modalParentViewController = self
        
        // Loading the ad (using IDs from the Constants class).
        banner.load(with: SASAdPlacement(siteId: Constants.siteID, pageId: pageId, formatId: Constants.bannerFormatID))
        
        return banner
    }
    
    func removeBanner(_ banner: SASBannerView?) {
        if (banner != nil) {
            banner?.modalParentViewController = nil
            banner?.delegate = nil
            banner?.removeFromSuperview()
        }
    }
    
    func bannerFrameForView(_ view: UIView, height: CGFloat) -> CGRect {
        // Convenient method to reset the frame of a banner
        return CGRect(x: 0, y: 0, width: view.frame.width, height: height)
    }
    
    // MARK: - Table View data source & delegate
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        var height: CGFloat = defaultCellHeight
        
        // If the cell is an ad cell, the actual cell height can be computed easily using the optimalAdViewHeightForContainer: method
        if (indexPath as NSIndexPath).row == banner1Row, let banner1Height = banner1?.optimalAdViewHeight(forContainer: tableView) {
            height = banner1Height
        } else if (indexPath as NSIndexPath).row == banner2Row, let banner2Height = banner2?.optimalAdViewHeight(forContainer: tableView) {
            height = banner2Height
        } else if (indexPath as NSIndexPath).row == banner3Row, let banner3Height = banner3?.optimalAdViewHeight(forContainer: tableView) {
            height = banner3Height
        }
        
        return height
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return numberOfCells
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if isAdCell(indexPath) {
            // Attach relevant banner to the container cell
            switch ((indexPath as NSIndexPath).row) {
            case banner1Row:
                return SASAdViewContainerCell(for: banner1!, in: tableView)
            case banner2Row:
                return SASAdViewContainerCell(for: banner2!, in: tableView)
            case banner3Row:
                return SASAdViewContainerCell(for: banner3!, in: tableView)
            default:
                NSLog("Error: cell should not be an AdCell")
                return tableView.dequeueReusableCell(withIdentifier: "ContentCell")!
            }
        } else if indexPath.row == 0 {
            // Configure description cell
            let reuseIdentifier = "DescriptionCell";
            var cell = tableView.dequeueReusableCell(withIdentifier:reuseIdentifier)
            if (cell == nil) {
                cell = UITableViewCell(style: UITableViewCell.CellStyle.default,
                                       reuseIdentifier: reuseIdentifier)
            }
            return cell!
        } else {
            // Configure regular cells
            let reuseIdentifier = "ContentCell";
            var cell = tableView.dequeueReusableCell(withIdentifier:reuseIdentifier)
            if (cell == nil) {
                cell = UITableViewCell(style: UITableViewCell.CellStyle.default,
                                       reuseIdentifier: reuseIdentifier)
            }
            let index:UILabel = cell?.viewWithTag(1) as! UILabel
            index.text = String(format: "%d", indexPath.row)
            index.layer.masksToBounds = true
            index.layer.cornerRadius = 15
            cell?.accessoryType = UITableViewCell.AccessoryType.disclosureIndicator
            return cell!
        }
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        //Deselect the cell if it's an ad to avoid messing with the ad creative.
        if (indexPath as NSIndexPath).row == banner1Row || banner2Row == (indexPath as NSIndexPath).row || banner3Row == (indexPath as NSIndexPath).row {
            //Ad Cell was selected, nothing to do here.
        } else {
            //Your normal behavior on cell click like pushing a new VC
            pushViewController()
        }
        tableView.deselectRow(at: indexPath, animated: false)
    }
    
    func pushViewController() {
        let controller = UIViewController()
        controller.navigationItem.title = "Details"
        controller.view.backgroundColor = view.backgroundColor
        let label = UILabel(frame: controller.view.bounds)
        label.autoresizingMask = [UIView.AutoresizingMask.flexibleWidth, UIView.AutoresizingMask.flexibleHeight]
        label.textAlignment = NSTextAlignment.center
        label.text = "Detail view controller"
        controller.view.addSubview(label)
        navigationController?.pushViewController(controller, animated: true)
    }
    
    func isAdCell(_ indexPath: IndexPath) -> Bool {
        // In this sample, ad cells are in fixed positions
        return (indexPath as NSIndexPath).row == banner1Row || (indexPath as NSIndexPath).row == banner2Row || banner3Row == (indexPath as NSIndexPath).row
    }
    
    func refreshAdCell(_ bannerView: SASBannerView?) {
        // Refresh the ad cell (or the whole table view if no ad cell is given in parameter)
        if bannerView == banner1 {
            tableView.reloadRows(at: [IndexPath(row: banner1Row, section: 0)], with: .automatic)
        } else if bannerView == banner2 {
            tableView.reloadRows(at: [IndexPath(row: banner2Row, section: 0)], with: .automatic)
        } else if bannerView == banner3 {
            tableView.reloadRows(at: [IndexPath(row: banner3Row, section: 0)], with: .automatic)
        } else {
            tableView.reloadData()
        }
    }
    
    // MARK: - Observe navigation bar hidden state to animate sticky view if needed
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if keyPath == "hidden" && stickyAdView != nil {
            if let navigationBar = self.navigationController?.navigationBar {
                if navigationBar.isHidden == false {
                    UIView.animate(withDuration: 0.3) {
                        if var newFrame = self.stickyAdView?.frame {
                            newFrame.origin.y = self.navigationBarTopOffset + navigationBar.frame.height
                            self.stickyAdView?.frame = newFrame
                        }
                    }
                } else {
                    UIView.animate(withDuration: 0.3) {
                        if var newFrame = self.stickyAdView?.frame {
                            newFrame.origin.y = self.navigationBarTopOffset
                            self.stickyAdView?.frame = newFrame
                        }
                    }
                }
            }
        }
    }
    
    // MARK: - SASBannerView delegate
    
    func bannerView(_ bannerView: SASBannerView, didDownloadAd ad: SASAd) {
        NSLog("An ad object has been loaded")
        
        // Refresh the table view when an ad is loaded
        refreshAdCell(bannerView)
    }
    
    func bannerViewDidLoad(_ bannerView: SASBannerView) {
        NSLog("Banner has been loaded")
    }
    
    func bannerView(_ bannerView: SASBannerView, didFailToLoadWithError error: Error) {
        NSLog("Banner has failed to load with error: \(error.localizedDescription)")
    }
    
    func bannerView(_ bannerView: SASBannerView, didSend videoEvent: SASVideoEvent) {
        // This delegate can be used to listen for events if the ad is a video.
        
        // For instance, you can use these events to check if the video has been played until the end
        // by listening to the event 'SASVideoEvent.Complete'
        if videoEvent == .complete {
            NSLog("The video has been played until the end")
        }
    }
    
    func bannerView(_ bannerView: SASBannerView, withStickyView stickyView: UIView, didStick stuck: Bool, withFrame stickyFrame: CGRect) {
        // This delegate can be used to be notified when a video read banner did stick or unstick, in this implementation as the navigation bar appears / disappears during vertical swipe we need to animate the sticky view
        
        if stuck {
            // keep a reference to the view to animate it if necessary
            self.stickyAdView = stickyView
        } else {
            self.stickyAdView = nil
        }
    }
    
}
